src/client/Microsoft.Identity.Client/ManagedIdentity/AppServiceManagedIdentitySource.cs

```diff
@@ -0,0 +1,88 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using Microsoft.Identity.Client.Internal;
+using Microsoft.Identity.Client.Utils;
+
+namespace Microsoft.Identity.Client.ManagedIdentity
+{
+    internal class AppServiceManagedIdentitySource : ManagedIdentitySource
+    {
+        // MSI Constants. Docs for MSI are available here https://docs.microsoft.com/azure/app-service/overview-managed-identity
+        protected string AppServiceMsiApiVersion = "2019-08-01";
+        protected string SecretHeaderName = "X-IDENTITY-HEADER";
+        protected string ClientIdHeaderName = "client_id";
+
+        private const string MsiEndpointInvalidUriError = "The environment variable MSI_ENDPOINT contains an invalid Uri.";
+
+        private readonly Uri _endpoint;
+        private readonly string _secret;
+        private readonly string _clientId;
+        private readonly string _resourceId;
+
+        public static ManagedIdentitySource TryCreate(RequestContext requestContext)
+        {
+            var msiSecret = EnvironmentVariables.IdentityHeader;
+            return TryValidateEnvVars(EnvironmentVariables.IdentityEndpoint, msiSecret, out Uri endpointUri)
+                ? new AppServiceManagedIdentitySource(requestContext, endpointUri, msiSecret)
+                : null;
+        }
+
+        protected static bool TryValidateEnvVars(string msiEndpoint, string secret, out Uri endpointUri)
+        {
+            endpointUri = null;
+
+            // if BOTH the env vars endpoint and secret values are null, this MSI provider is unavailable.
```

Comment:
would it be a good idea to list those env variables here in the comment?

Reply:
I don't think so. It only explains why we are not creating this source as managed identity.